### PR TITLE
Add support for Windows ARM64

### PR DESCRIPTION
**Pull request by RockLakeGrass**
I have a prebuilt file compiled for ARM64 and I don't know how to upload it, so it's temporarily stored in my repository

### CODE REVIEW

**Comment by Starbuck5** [comment]
Hey, I'm quite impressed with your effort on this issue.

I'm not sure why you closed #3276 already.

I think the goal to improve this PR is to minimize the code changes needed, and I have a few suggestions.

Firstly, for SCALE_MMX. See this bit of code: https://github.com/pygame/pygame/blob/main/src_c/scale.h#L32-L36. I think changing this so it just isn't defined on Windows ARM64 would be better than dealing with it further down the line throughout the code.

Secondly, I don't want a custom SETUP file. We already have three, and it's a pain when you have to change one and guess if changing the others will work the same way. I think what you should do is set up a custom SDL arm prebuilt repo that emulates the main SDL prebuilt repo. That way you can use the normal SETUP generation process, just switch out the URLs in the buildconfig to download from the new repo on Windows ARM64.

FYI: Since you haven't previously committed to this repository, Github won't run tests without member approval. That's why your tests may take a while to run. Since one of us has to press the button

FYI2: The best resource for pygame contributors is the `contributing` channel on the pygame discord (https://discord.com/invite/ZuB2RySPRJ), that's the best place to get help from other contributors.

**Comment by RockLakeGrass** [comment]
> I'm not sure why you closed https://github.com/pygame/pygame/pull/3276 already.

I think I've finished the transplant. Do I need to reopen it?

> for SCALE_MMX. See this bit of code

I'll try it later, but I think my solution is the fastest way to get pyGame to support win-arm.

> I don't want a custom SETUP file

Secondly, Since the Python official MSVC module does not support win-arm64, I can only customize the setup file. I can support win arm by modifying the code of MSVC module, but I don't think this is a good method.

> the channel on the pygame discord 

At last, I'm sorry I can't open Pygame discord

**Comment by Starbuck5** [comment]
Oh sorry, wrong issue tag. I meant #3267.

That issue is titled "Add support for Windows on ARM," and that's what this PR does. So I don't see why you would close the issue before the PR is merged.

> I'll try it later, but I think my solution is the fastest way to get pyGame to support win-arm.

I care more about maintainability than convenience of implementation, but I'm confident this will be more convenient to implement as well, with less code change.

> Secondly, Since the Python official MSVC module does not support win-arm64, I can only customize the setup file. I can support win arm by modifying the code of MSVC module, but I don't think this is a good method.

I don't see the relevance of MSVC to what you're talking about. It's just two different ways of generating the SETUP file. I don't want to do it the way you have it currently.

> At last, I'm sorry I can't open Pygame discord

"can't" ? If you don't have discord and don't want to get it, no worries, I just wanted to make sure you knew the deal.

**Comment by Starbuck5** [comment]
The other reason to make a generic SDL Windows ARM prebuilt repo is because other people will be able to use it. Maybe other SDL2 developers could use that, or other libraries that depend on SDL2.

**Comment by RockLakeGrass** [comment]
> So I don't see why you would close the issue before the PR is merged.

I've reopened it.

>  don't see the relevance of MSVC to what you're talking about

```Traceback (most recent call last):
  File "C:\Users\m1781\Downloads\pygame-2.1.2\pygame-2.1.2\setup.py", line 359, in <module>
    buildconfig.config.main(AUTO_CONFIG)
  File "C:\Users\m1781\Downloads\pygame-2.1.2\pygame-2.1.2\buildconfig\config.py", line 225, in main
    deps = CFG.main(**kwds)
  File "C:\Users\m1781\Downloads\pygame-2.1.2\pygame-2.1.2\buildconfig\config_win.py", line 511, in main
    return setup_prebuilt_sdl2(prebuilt_dir)
  File "C:\Users\m1781\Downloads\pygame-2.1.2\pygame-2.1.2\buildconfig\config_win.py", line 471, in setup_prebuilt_sdl2
    DEPS.configure()
  File "C:\Users\m1781\Downloads\pygame-2.1.2\pygame-2.1.2\buildconfig\config_win.py", line 336, in configure
    from . import vstools
  File "C:\Users\m1781\Downloads\pygame-2.1.2\pygame-2.1.2\buildconfig\vstools.py", line 11, in <module>
    compiler.initialize()
  File "C:\Complier\Python310-arm64\lib\distutils\msvc9compiler.py", line 347, in initialize
    raise DistutilsPlatformError("--plat-name must be one of %s" %
distutils.errors.DistutilsPlatformError: --plat-name must be one of ('win32', 'win-amd64')
```
During the generation of the Setup file, the msvc9compiler module was called, which does not support win-arm64.

> "can't" ? 

It seems that due to national policies, this website is blocked, and the use of VPN in our country is not allowed.

> The other reason to make a generic SDL Windows ARM prebuilt repo is because other people will be able to use it. Maybe other SDL2 developers could use that, or other libraries that depend on SDL2.

I don't understand what you're saying. Can't I prebuilt the SDL?

**Comment by RockLakeGrass** [comment]
> See this bit of code: https://github.com/pygame/pygame/blob/main/src_c/scale.h#L32-L36.

The "scale.h" has been modified to suit your suggestion.It's really easier in this way.

File: src_c/alphablit.c
Comment (Starbuck5) [inline src_c/alphablit.c:269]: What happens if you don't add this?

**Comment by Starbuck5** [comment]
> I don't understand what you're saying. Can't I prebuilt the SDL?

So many of your SETUP changes are to accommodate the fact that the prebuilts now live in a folder called prebuilt-ARM64. But it doesn't have to be like that. You could package the prebuilts so they look like pygame is expecting, just at a different website. And that way it would also be helpful to other people trying to use Windows ARM. For example, the SDL_image releases site: https://www.libsdl.org/projects/SDL_image/release/.

> During the generation of the Setup file, the msvc9compiler module was called, which does not support win-arm64.

Ugh. I'd prefer not to have another SETUP file in there, there must be a better way to do this.

File: src_c/alphablit.c
Comment (RockLakeGrass) [reply src_c/alphablit.c:269]: If you don't, he will get a LINK error and will not be able to parse the symbols.
Because these symbols cannot be used on Windows arm.

Comment (Starbuck5) [reply src_c/alphablit.c:269]: I'm curious what symbols?

**Comment by RockLakeGrass** [comment]
I don't know why there was an error on Ubuntu after modifying "scale.h". In that case, I had to change the code back to the previous version, at least that's correct.

> just at a different website

First, I don't have permission to make changes to the SDL's publishing file. Secondly, since I need to upload the SDL of ARM64, why don't I just pack it into one file for Pygame to download? Then, I made some changes to Pygame in the SDL package, such as generating lib files, which may not be allowed officially in SDL

> there must be a better way to do this

See this bit of code https://github.com/pygame/pygame/blob/main/buildconfig/config_win.py#L420. If you don't customize the SETUP file, you will get an error on this line, which I think is unavoidable during the automatic generation of the SETUP file.

File: src_c/alphablit.c
Comment (RockLakeGrass) [reply src_c/alphablit.c:269]: Some symbols about RGB functions, such as blit_blend_rgb_sub_sse2.

After my analysis, the reasons for incompatibility are as follows: First, MSVC ARM64 does not support SSE, which is a fact. In this file, there is already a way to be compatible with ARM64, even with sse2neon, in the GitHub repository of this header file, I found an (https://github.com/DLTcollab/sse2neon/issues/384), which means that this header file does not support MSVC.

**Comment by RockLakeGrass** [comment]
> Ubuntu sdist / build (ubuntu-22.04) (pull_request) Failing after 29s — build (ubuntu-22.04)

I don't know why it's so amazing that it doesn't go wrong with the same code before

**Comment by Starbuck5** [comment]
That fails every time.

**Comment by RockLakeGrass** [comment]
https://github.com/pygame/pygame/pull/3278/commits/a5b27aa625f2f3cdd5e438593b46c575d00f4aba
This is the right time.

**Comment by RockLakeGrass** [comment]
SDL components in prebuilt files have been downloaded as separate files, and the reason why the latest version of SDL was used is that libmpg123 was discarded in the latest version, which allows SDLmixer to be fully ported to Windows ARM64.

**Comment by RockLakeGrass** [comment]
From https://ci.appveyor.com/project/pygame/pygame/builds/44099669/job/92rn46mco7ow5rxi
```
Installing Python 3.10.0 64 to C:\Python310-x64...
Downloading https://www.python.org/ftp/python/3.10.0/python-3.10.0-amd64.exe...
Exception calling "DownloadFile" with "2" argument(s): "An exception occurred during a WebClient request."
At line:12 char:5
+     (New-Object Net.WebClient).DownloadFile($downloadUrl, $exePath)
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : WebException
 
Installing...
```
How can there be problems when downloading Python from this website

File: src_c/alphablit.c
Comment (ankith26) [inline src_c/alphablit.c:313]: I feel these checks can be removed.

In theory (untested), the neon path (neon being ARM SIMD) should also work on ARM64 windows.

In this same file, we have
```c
#if !defined(PG_ENABLE_ARM_NEON) && defined(__aarch64__)
// arm64 has neon optimisations enabled by default, even when fpu=neon is not
// passed
#define PG_ENABLE_ARM_NEON 1
#endif
```
somewhere at the top. I doubt `__aarch64__` is defined on MSVC, so I propose that this should become
```c
#if !defined(PG_ENABLE_ARM_NEON) && (defined(__aarch64__) || defined(_M_ARM64))
// arm64 has neon optimisations enabled by default, even when fpu=neon is not
// passed
#define PG_ENABLE_ARM_NEON 1
#endif
```

File: src_c/transform.c
Comment (ankith26) [inline src_c/transform.c:1614]: I think (but not sure) that this should not need special-casing.

**Review by ankith26 — COMMENTED** [review]
Gotta say, really impressive for a first time PR, so thanks!

PS: This PR might take a while to review

**Comment by Starbuck5** [comment]
Wow, big changes since I last looked! I like it.

However, I'm surprised you went back from changing the definition of SCALE_MMX_SUPPORT (https://github.com/pygame/pygame/blob/main/src_c/scale.h#L32-L36) like you did before. That seems more elegant?

File: src_c/alphablit.c
Comment (RockLakeGrass) [reply src_c/alphablit.c:313]: https://github.com/DLTcollab/sse2neon/issues/384
Sse2neon does not support MSVC.

**Comment by RockLakeGrass** [comment]
> However, I'm surprised you went back from changing the definition of SCALE_MMX_SUPPORT (https://github.com/pygame/pygame/blob/main/src_c/scale.h#L32-L36) like you did before. That seems more elegant?

If you write like last time, you cannot pass format-lint-code-check.

I see your submission, which updates the SDL, and I want to wait until SDL releases the latest official version and you update all SDL components before I migrate the latest version of SDL. Because Windows ARM64 is still using the preview version.

**Comment by Starbuck5** [comment]
> it cannot pass format-lint-code-check.

Run it through clang-format and it will format it so it passes.

**Review by ankith26 — COMMENTED** [review]
The code looks fine on a skim through.

I see you have put up your own SDL ARM64 releases, this is impressive work, I'd prefer to take this a step further and upstream this (ie hopefully get official SDL support), doing this would be helpful for more SDL users, and would also centralise the maintenance efforts needed to continue supporting compiling on ARM Windows.

Thanks again for contributing!

**Comment by RockLakeGrass** [comment]
>  I'd prefer to take this a step further and upstream this (ie hopefully get official SDL support)

I made this [request](https://github.com/libsdl-org/SDL/issues/5882) in the SDL repository, but in my opinion, the official will not support Windows arm, and even if it does, it will only be supported in the new version. After testing, the latest sdl_image cannot be used in Pygame (it can be compiled but cannot pass the test)

**Review by illume — APPROVED** [review]
👍 thank you very much.

I'm going to continue this in another branch / PR.

**Pull request merged**
